### PR TITLE
gRPC smserver  with new api

### PR DESCRIPTION
**Pull request by Eabramov84**

### CODE REVIEW

File: smserver/smserver_test.go [hunk 1]
```diff
@@ -515,3 +661,89 @@ func (client *testClient) close() {
...
+	for i := range expected {
+		if expected[i].LogId != received[i].LogID {
+			return aoserrors.New("incorrect LogID")
+		}
+
+		if received[i].From == nil {
+			if expected[i].From != nil {
+				return aoserrors.New("incorrect from timestamp")
+			}
+		} else if *received[i].From != expected[i].From.AsTime() {
```
Comment (al1img) [inline smserver/smserver_test.go:1328]: It will crash if expected[i].From == nil.

File: smserver/smserver_test.go [hunk 2]
```diff
@@ -515,3 +661,89 @@ func (client *testClient) close() {
...
+			return aoserrors.New("incorrect from timestamp")
+		}
+
+		if received[i].Till == nil {
+			if expected[i].Till != nil {
+				return aoserrors.New("incorrect till timestamp")
+			}
+		} else if *received[i].Till != expected[i].Till.AsTime() {
```
Comment (al1img) [inline smserver/smserver_test.go:1336]: Ditto.

File: smserver/smserver_test.go [hunk 3]
```diff
@@ -515,3 +661,89 @@ func (client *testClient) close() {
...
+			return aoserrors.New("incorrect till timestamp")
+		}
+
+		if received[i].ServiceID != expected[i].Instance.ServiceId {
+			return aoserrors.New("incorrect ServiceID")
+		}
+
+		if received[i].SubjectID == nil {
+			if expected[i].Instance.SubjectId != "" {
+				return aoserrors.New("incorrect subject id")
```
Comment (al1img) [inline smserver/smserver_test.go:1346]: id -> ID

File: smserver/smserver_test.go [hunk 4]
```diff
@@ -515,3 +661,89 @@ func (client *testClient) close() {
...
+			}
+		} else if *received[i].SubjectID != expected[i].Instance.SubjectId {
+			return aoserrors.New("incorrect subject id")
```
Comment (al1img) [inline smserver/smserver_test.go:1349]: id  -> ID

File: smserver/smserver.go [hunk 5]
```diff
@@ -421,3 +432,102 @@ func cloudprotocolLogToPb(log cloudprotocol.PushLog) (pbLog *pb.LogData) {
...
 	return pbLog
 }
+
+// nolint:funlen
+func cloudprotocolAlertToPB(alert *cloudprotocol.AlertItem) (pbAlert *pb.Alert, err error) {
+	pbAlert = &pb.Alert{Tag: alert.Tag, Timestamp: timestamppb.New(alert.Timestamp)}
+
+	switch alert.Tag {
+	case cloudprotocol.AlertTagSystemError:
+		sysAlert, ok := alert.Payload.(cloudprotocol.SystemAlert)
```
Comment (al1img) [inline smserver/smserver.go:625]: create separate functions for each alert tag

File: smserver/smserver.go [hunk 6]
```diff
@@ -421,3 +432,102 @@ func cloudprotocolLogToPb(log cloudprotocol.PushLog) (pbLog *pb.LogData) {
...
+		if !ok {
+			return nil, errIncorrectAlertType
```
Comment (al1img) [inline smserver/smserver.go:627]: aoserrors.Wrap

File: smserver/smserver.go [hunk 7]
```diff
@@ -421,3 +432,102 @@ func cloudprotocolLogToPb(log cloudprotocol.PushLog) (pbLog *pb.LogData) {
...
+		}
+
+		pbAlert.Payload = &pb.Alert_SystemAlert{SystemAlert: &pb.SystemAlert{Message: sysAlert.Message}}
+
+	case cloudprotocol.AlertTagAosCore:
+		coreAlert, ok := alert.Payload.(cloudprotocol.CoreAlert)
+		if !ok {
+			return nil, errIncorrectAlertType
```
Comment (al1img) [inline smserver/smserver.go:635]: aoserrors.Wrap

File: smserver/smserver.go [hunk 8]
```diff
@@ -421,3 +432,102 @@ func cloudprotocolLogToPb(log cloudprotocol.PushLog) (pbLog *pb.LogData) {
...
+
+		pbResAlert := pb.ResourceValidateAlert{}
+
+		for _, resAlertData := range resAlert.ResourcesErrors {
+			pbResAlertElement := pb.ResourceValidateErrors{
+				Name:     resAlertData.Name,
+				ErrorMsg: make([]string, len(resAlertData.Errors)),
+			}
+
+			_ = copy(pbResAlertElement.ErrorMsg, resAlertData.Errors)
```
Comment (al1img) [inline smserver/smserver.go:657]: Why do you need to copy it?

File: smserver/smserver.go [hunk 9]
```diff
@@ -421,3 +432,102 @@ func cloudprotocolLogToPb(log cloudprotocol.PushLog) (pbLog *pb.LogData) {
...
+			Instance:   cloudprotocolInstanceIdentToPb(instAlert.InstanceIdent),
+			AosVersion: instAlert.AosVersion,
+			Message:    instAlert.Message,
+		}}
+	}
+
+	return pbAlert, nil
+}
+
+func cloudprotocolInstanceIdentToPb(ident cloudprotocol.InstanceIdent) *pb.InstanceIdent {
```
Comment (al1img) [inline smserver/smserver.go:714]: Pb -> PB

File: smserver/smserver.go [hunk 10]
```diff
@@ -44,6 +46,8 @@ import (
  * Vars
  ******************************************************************************/
```
Comment (al1img) [inline smserver/smserver.go]: Extend headers up to 120 symbols.

File: smserver/smserver.go [hunk 11]
```diff
@@ -357,3 +381,43 @@ func (server *SMServer) handleChannels() {
 		}
 	}
 }
+
+func getFromTillTimeFromPb(fromPb, tillPb *timestamp.Timestamp) (from, till *time.Time) {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply smserver/smserver.go:565]: done

File: smserver/smserver_test.go [hunk 12]
```diff
@@ -515,3 +661,89 @@ func (client *testClient) close() {
 		client.connection.Close()
 	}
 }
+
+func checkLogs(notification pb.SMService_SubscribeSMNotificationsClient, testLogs []testLogData) (err error) {
+	notificationChan := make(chan *pb.SMNotifications, 1)
+
+	go func() {
+		for {
+			smData, _ := notification.Recv()
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply smserver/smserver_test.go:1283]: added

File: smserver/smserver_test.go [hunk 13]
```diff
@@ -515,3 +661,89 @@ func (client *testClient) close() {
...
+		}
+	}()
+
+	var currentIndex int
+
+	for {
+		select {
+		case rawLog := <-notificationChan:
+			logData := rawLog.GetLog()
+			if logData == nil {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply smserver/smserver_test.go:1294]: done

File: smserver/smserver_test.go [hunk 14]
```diff
@@ -515,3 +661,89 @@ func (client *testClient) close() {
...
+				return errIncorrectType
+			}
+
+			if !proto.Equal(logData, &testLogs[currentIndex].expectedPbLog) {
+				return aoserrors.New("received log != send log")
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply smserver/smserver_test.go:1299]: done

File: smserver/smserver_test.go [hunk 15]
```diff
@@ -515,3 +661,89 @@ func (client *testClient) close() {
...
+	}
+}
+
+func compareServiceLogRequests(expected []pb.InstanceLogRequest, received []cloudprotocol.RequestServiceLog) error {
+	if len(expected) != len(received) {
+		return aoserrors.New("incorrect len of requests")
+	}
+
+	for i := range expected {
+		if expected[i].LogId != received[i].LogID {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply smserver/smserver_test.go:1320]: done

File: smserver/smserver_test.go (hunk 1 above)
Comment (Eabramov84) [reply smserver/smserver_test.go:1328]: fixed

File: smserver/smserver_test.go (hunk 3 above)
Comment (Eabramov84) [reply smserver/smserver_test.go:1346]: fixed

File: smserver/smserver.go (hunk 5 above)
Comment (Eabramov84) [reply smserver/smserver.go:625]: done

File: smserver/smserver.go (hunk 8 above)
Comment (Eabramov84) [reply smserver/smserver.go:657]: removed

File: smserver/smserver.go (hunk 10 above)
Comment (Eabramov84) [reply smserver/smserver.go]: did in separate commit

File: smserver/smserver_test.go [hunk 16]
```diff
@@ -73,6 +81,20 @@ type testResourceManager struct {
...
+	intrenalLog   cloudprotocol.PushLog
+	expectedPbLog pb.LogData
+}
+
+/***********************************************************************************************************************
+ * Vars
+ **********************************************************************************************************************/
+
+var (
+	errTimeout       = errors.New("timeout")
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply smserver/smserver_test.go:129]: removed

File: smserver/smserver.go [hunk 17]
```diff
@@ -232,26 +232,28 @@ func (server *SMServer) GetAllStatus(ctx context.Context, req *empty.Empty) (sta
...
+	if err := server.boardConfigProcessor.CheckBoardConfig(boardConfig.GetBoardConfig()); err != nil {
+		return &emptypb.Empty{}, aoserrors.Wrap(err)
+	}
 
-	return &pb.BoardConfigStatus{VendorVersion: version}, err
+	return &emptypb.Empty{}, nil
 }
 
 // SetBoardConfig sets new board configuration.
 func (server *SMServer) SetBoardConfig(ctx context.Context, boardConfig *pb.BoardConfig) (ret *empty.Empty, err error) {
```
Comment (MykolaSolyanko) [inline smserver/smserver.go:239]: in CheckBoardConfig you are not using the name returned value, maybe here also remove it?

File: smserver/smserver.go [hunk 18]
```diff
@@ -259,13 +233,37 @@ func (server *SMServer) SetBoardConfig(ctx context.Context, boardConfig *pb.Boar
...
+	}
+
+	return &emptypb.Empty{}, nil
 }
 
-// InstallService removes aos service.
-func (server *SMServer) RemoveService(ctx context.Context, service *pb.RemoveServiceRequest) (ret *empty.Empty, err error) {
-	return &emptypb.Empty{}, server.launcher.UninstallService(service)
+// GetServicesStatus gets installed services info.
+func (server *SMServer) GetServicesStatus(context.Context, *empty.Empty) (services *pb.ServicesStatus, err error) {
```
Comment (MykolaSolyanko) [inline smserver/smserver.go:268]: maybe here the names of the returned values are redundant?

File: smserver/smserver_test.go [hunk 19]
```diff
@@ -95,6 +101,12 @@ type testLogData struct {
 	expectedPBLog pb.LogData
 }
 
+type testServiceInstallRequeststruct struct {
```
Comment (MykolaSolyanko) [inline smserver/smserver_test.go:112]: `testServiceInstallRequest` ?

File: smserver/smserver_test.go [hunk 20]
```diff
@@ -725,40 +736,83 @@ func TestServiceStateProcessing(t *testing.T) {
...
-		log.Error("received state request data != sent data")
+	if _, err := smServer.SetInstanceState(context.Background(), &pb.InstanceState{
+		Instance:      &pb.InstanceIdent{ServiceId: "service1", SubjectId: "s1", Instance: 2},
+		StateChecksum: checkSumStr,
+		State:         []byte(stateStr),
+	}); err != nil {
+		t.Fatalf("Can't set instance state: %s", err)
 	}
+
+	reflect.DeepEqual(expectedUpdateState, stateHandler.currentUpdateState)
```
Comment (MykolaSolyanko) [inline smserver/smserver_test.go:967]: should be checked? and I can't find where currentUpdateState is set

File: smserver/smserver_test.go [hunk 21]
```diff
@@ -207,6 +211,160 @@ func TestBoardConfiguration(t *testing.T) {
...
+
+		sendEnvVars.EnvVars = append(sendEnvVars.EnvVars, &pb.OverrideInstanceEnvVar{
+			Instance: &pb.InstanceIdent{ServiceId: "id" + strconv.Itoa(i), SubjectId: "s1", Instance: int64(i - 1)},
+			Vars: []*pb.EnvVarInfo{
+				{VarId: "varId" + strconv.Itoa(i), Variable: "variable " + strconv.Itoa(i)},
+				{VarId: "varId2" + strconv.Itoa(i), Variable: "variable 2" + strconv.Itoa(i), Ttl: timestamppb.New(curentTime)},
+			},
+		})
+
+		sendEnvVars.EnvVars[0].Instance.SubjectId = ""
```
Comment (MykolaSolyanko) [inline smserver/smserver_test.go:251]: you did the same after the loop, it's ok?

File: smserver/smserver_test.go [hunk 22]
```diff
@@ -207,6 +211,160 @@ func TestBoardConfiguration(t *testing.T) {
...
+	expectedRequest[0].InstanceFilter.SubjectID = nil
+	expectedStatus.EnvVarsStatus[0].Instance.SubjectId = ""
+
+	envVarStatus, err := smServer.OverrideEnvVars(context.Background(), sendEnvVars)
+	if err != nil {
+		t.Fatalf("Can't override env vars: %s", err)
+	}
+
+	if !proto.Equal(envVarStatus, expectedStatus) {
+		t.Errorf("Incorrect overrode env var status")
```
Comment (MykolaSolyanko) [inline smserver/smserver_test.go:279]: overrode -> override ?

File: smserver/smserver_test.go [hunk 23]
```diff
@@ -207,6 +211,160 @@ func TestBoardConfiguration(t *testing.T) {
...
+	}
+
+	if reflect.DeepEqual(expectedRequest, testLauncher.envVarRequest) {
```
Comment (MykolaSolyanko) [inline smserver/smserver_test.go:282]: a little confusing this check, we created the variable `expectedRequest` but in fact we have the opposite. But up to you

File: smserver/smserver_test.go [hunk 24]
```diff
@@ -418,6 +440,101 @@ func TestServiceStateProcessing(t *testing.T) {
...
+
+	logProvider := testLogProvider{channel: make(chan cloudprotocol.PushLog)}
+
+	smServer, err := smserver.New(&smConfig, nil, nil, nil, nil, nil, &logProvider, nil, nil, true)
+	if err != nil {
+		t.Fatalf("Can't create SM server: %s", err)
+	}
+
+	go func() {
+		if err := smServer.Start(); err != nil {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply smserver/smserver_test.go:983]: done on separate commit

File: smserver/smserver_test.go [hunk 25]
```diff
@@ -219,92 +220,205 @@ func TestAlertNotifications(t *testing.T) {
...
-		t.Errorf("Can't receive validation alert: %s", err)
-	}
+	select {
+	case <-alertChan:
+		t.Error("Unexpected Alert")
 
-	if !proto.Equal(receivedValidationAlert.GetAlert(), validationAlert) {
-		log.Error("received validation alert != send alert")
+	case <-time.After(2 * time.Second):
+		break
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply smserver/smserver_test.go:752]: removed

File: smserver/smserver.go [hunk 26]
```diff
@@ -285,7 +287,37 @@ func (server *SMServer) OverrideEnvVars(ctx context.Context,
...
+
+	if err = server.layerProvider.InstallLayer(installInfo, layer.Url, fileInfo); err != nil {
+		return &emptypb.Empty{}, aoserrors.Wrap(err)
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+// GetLayersStatus gets installed layer info.
+func (server *SMServer) GetLayersStatus(context.Context, *empty.Empty) (layers *pb.LayersStatus, err error) {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply smserver/smserver.go]: removed

File: smserver/smserver_test.go (hunk 19 above)
Comment (Eabramov84) [reply smserver/smserver_test.go:112]: renamed

File: smserver/smserver_test.go (hunk 21 above)
Comment (Eabramov84) [reply smserver/smserver_test.go:251]: removed

File: smserver/smserver_test.go (hunk 20 above)
Comment (Eabramov84) [reply smserver/smserver_test.go]: fixed

File: smserver/smserver_test.go (hunk 22 above)
Comment (Eabramov84) [reply smserver/smserver_test.go:279]: fixed

File: smserver/smserver_test.go (hunk 23 above)
Comment (Eabramov84) [reply smserver/smserver_test.go:282]: tests also performs validation of expected request

**Review by MykolaSolyanko — APPROVED** [review]

File: smserver/smserver_test.go [hunk 27]
```diff
@@ -195,6 +197,71 @@ func TestConnection(t *testing.T) {
...
+		t.Fatalf("Can't set board config: %s", err)
+	}
+
+	if !proto.Equal(expectedBCStatus, bcStatus) {
+		t.Error("Incorrect board config status")
+	}
+
+	if _, err := client.pbclient.CheckBoardConfig(
+		context.Background(), &pb.BoardConfig{BoardConfig: expectedBoardConfig}); err != nil {
+		t.Fatalf("Can't set board config: %s", err)
```
Comment (al1img) [inline smserver/smserver_test.go]: Check board config error. etc.

File: smserver/smserver_test.go [hunk 28]
```diff
@@ -186,14 +198,95 @@ func TestBoardConfiguration(t *testing.T) {
...
+				Size: uint64(i + 500),
+			},
+		}
+
+		expectedStatus := pb.ServiceStatus{
+			ServiceId: "testService" + strconv.Itoa(i), AosVersion: uint64(i),
+		}
+
+		if _, err := smServer.InstallService(context.Background(), &installRequest); err != nil {
+			t.Fatalf("Can't install service")
```
Comment (al1img) [inline smserver/smserver_test.go:402]: add %s, err

File: smserver/smserver_test.go [hunk 29]
```diff
@@ -719,39 +730,84 @@ func TestServiceStateProcessing(t *testing.T) {
...
 
-	receivedSateRequest, err := notifications.Recv()
+	receivedStateRequest, err := notifications.Recv()
 	if err != nil {
-		t.Errorf("Can't receive monitoring data: %s", err)
+		t.Fatalf("Can't receive notification data: %s", err)
+	}
+
+	if !proto.Equal(expectedStateRequest, receivedStateRequest) {
+		log.Error("received state request != sent data")
```
Comment (al1img) [inline smserver/smserver_test.go:903]: t.Error

File: smserver/smserver.go [hunk 30]
```diff
@@ -370,6 +425,13 @@ func (server *SMServer) GetInstanceCrashLog(ctx context.Context, req *pb.Instanc
 func (server *SMServer) handleChannels() {
 	for {
 		select {
+		case runtimeStatus := <-server.runtimeStatusChannel:
+			if err := server.sendRuntimeInstanceNotifications(runtimeStatus); err != nil {
+				log.Error(err)
```
Comment (al1img) [inline smserver/smserver.go:437]: Display error message

File: smserver/smserver_test.go [hunk 31]
```diff
@@ -204,6 +208,158 @@ func TestBoardConfiguration(t *testing.T) {
...
+	expectedRequest[0].InstanceFilter.SubjectID = nil
+	expectedStatus.EnvVarsStatus[0].Instance.SubjectId = ""
+
+	envVarStatus, err := smServer.OverrideEnvVars(context.Background(), sendEnvVars)
+	if err != nil {
+		t.Fatalf("Can't override env vars: %s", err)
+	}
+
+	if !proto.Equal(envVarStatus, expectedStatus) {
+		t.Errorf("Incorrect override env var status")
```
Comment (al1img) [inline smserver/smserver_test.go:264]: t.Error

File: smserver/smserver_test.go [hunk 32]
```diff
@@ -204,6 +208,158 @@ func TestBoardConfiguration(t *testing.T) {
...
+	if !reflect.DeepEqual(expectedRunRequest, testLauncher.runRequest) {
+		t.Errorf("Incorrect run instances request")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	notifications, err := client.pbclient.SubscribeSMNotifications(ctx, &emptypb.Empty{})
+	if err != nil {
+		t.Fatalf("Can't subscribes: %s", err)
```
Comment (al1img) [inline smserver/smserver_test.go:321]: Can't subscribe

File: servicemanager/servicemanager.go [hunk 33]
```diff
@@ -125,14 +125,20 @@ func New(config *config.Config, serviceInfoProvider ServiceStorage) (sm *Service
```
Comment (al1img) [inline servicemanager/servicemanager.go:125]: Fix commit message: implementation

File: smserver/smserver_test.go (hunk 27 above)
Comment (Eabramov84) [reply smserver/smserver_test.go]: fixed

File: smserver/smserver_test.go (hunk 29 above)
Comment (Eabramov84) [reply smserver/smserver_test.go:903]: fixed

File: smserver/smserver_test.go (hunk 31 above)
Comment (Eabramov84) [reply smserver/smserver_test.go:264]: done

File: servicemanager/servicemanager_test.go [hunk 34]
```diff
@@ -304,6 +305,15 @@ func TestImageParts(t *testing.T) {
 		t.Errorf("Can't get service info: %s", err)
 	}
 
+	services, err := sm.GetAllServicesStatus()
+	if err != nil {
+		t.Errorf("Can't get all services")
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply servicemanager/servicemanager_test.go:310]: done

File: smserver/smserver.go [hunk 35]
```diff
@@ -54,13 +55,15 @@ var errIncorrectAlertType = errors.New("incorrect alert type")
  * Types
  ******************************************************************************/
 
-// ServiceLauncher services launcher interface
```
Comment (al1img) [inline smserver/smserver.go:48]: Return comment back.

**Review by al1img — APPROVED** [review]

**Pull request merged**